hooks/useGovernanceUnderlyingTokenAccounts.ts

```diff
@@ -0,0 +1,110 @@
+import BigNumber from 'bignumber.js'
+import { useCallback, useEffect, useState } from 'react'
+import { PublicKey } from '@solana/web3.js'
+import { getOwnedTokenAccounts, tryGetMint } from '@utils/tokens'
+import useWalletStore from 'stores/useWalletStore'
+import { SPL_TOKENS } from '@utils/splTokens'
+import { BN } from '@project-serum/anchor'
+import { findATAAddrSync } from '@uxdprotocol/uxd-client'
+
+export type OwnedTokenAccountInfo = {
+  pubkey: PublicKey
+  mint: PublicKey
+  amount: BN
+  uiAmount: number
+  mintDecimals: number
+  mintName: string
+
+  // could be either the Associated Token Account of the mint
+  // or a Token Account
+  isATA: boolean
+}
+
+export type OwnedTokenAccountInfos = OwnedTokenAccountInfo[]
+
+// Loads all the token accounts related to the governance public key
+export default function useGovernanceUnderlyingTokenAccounts(
+  governancePk: PublicKey | null
+) {
+  const connection = useWalletStore((state) => state.connection)
+
+  const [
+    ownedTokenAccounts,
+    setOwnedTokenAccounts,
+  ] = useState<OwnedTokenAccountInfos | null>(null)
+
+  const getOwnedTokenAccountsFn = useCallback(async () => {
+    if (!connection || !governancePk) return null
+
+    const accounts = await getOwnedTokenAccounts(
+      connection.current,
+      governancePk
+    )
+
+    const ownedTokenAccountInfos = accounts.map((x) => ({
+      pubkey: x.publicKey,
+      mint: x.account.mint,
+      amount: new BN(x.account.amount.toString()),
+    }))
+
+    // Compute a list of mints related to token accounts
+    const mints = Array.from(
+      ownedTokenAccountInfos.reduce((tmp, x) => tmp.add(x.mint), new Set())
```

Comment:
looping over `accounts` once, then once again through `ownedTokenAccountInfos` here

Reply:
Also, why create a `Set` here ? Many loops wow

Reply:
Set in order to not call `tryGetMint` twice for the same mint

Reply:
Maybe extract the mints early, dedupe using Set spread pattern, and then hold a map `{ [mint]: accounts }` ?